internal/app/upstream/multiplexer.go

```diff
@@ -0,0 +1,37 @@
+package multiplexer
+
+import (
+	"context"
+
+	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"google.golang.org/grpc"
+)
+
+// Multiplexer handles the connections to the origin server
+type Multiplexer interface {
+	// QueueRequest starts a stream with the origin server
+	// All incoming discovery requests arrive on the request channel
+	// All responses go out through the response channel
+	// All irrecoverable errors go out through the error channel
+	QueueRequest(context.Context, chan *v2.DiscoveryRequest, chan *v2.DiscoveryResponse, chan error)
```

Comment:
Does this need to return a handle or some way of cleaning up? How do things get removed from the multiplexer?

Reply:
The context.cancel should be able to perform the cleanups, it will be part of implementation.